### PR TITLE
feature(aws_vpc_peering): switch multi DC cases to work with vpc-peering

### PR DESCRIPTION
**Pull request by fruch**
- [x] Recreate our VPC configuration so we can do peering across multiple regions
- [x] Fix multi-dc test cases to be operational 
- [x] make sure external ports for SCT SGs are closed
- [x] automate Jenkins workers creation
- [x] add license headers to new files
- [x] go over all multi-dc test cases and make sure they aren't using public ip in their configuration 
- [x] create a way to connect/open ports as needed (SG per test ?, bastion?)
- [x] test multi-dc case on GCP, to check it's not broken by those changes
- [x] check if `longevity-200gb-48h-network` is broken cause of this change
- [x] update the Readme with instruction how to use new hydra commands
- [x] test [artifacts-azure-image.jenkinsfile
- [x] check when aws-provision test is failing
- [x] check that the manger ipv6 isn't broken by those changes (test working, multi-dc ipv6 isn't, not related to vpc-peering)
- [x] tested a k8s job with this branch (job)[https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/longevity-scylla-operator-3h-eks/5/]

Trello card: https://trello.com/c/j8iS25lt

## PR pre-checks (self review)

- [ ] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [ ] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

File: vars/getJenkinsLabels.groovy [hunk 1]
```diff
@@ -9,11 +9,11 @@ def call(String backend, String region=null, String datacenter=null, String loca
 
     }
 
-    def jenkins_labels = ['aws-eu-west-1': 'aws-sct-builders-eu-west-1-new',
-                          'aws-eu-west-2': 'aws-sct-builders-eu-west-2',
-                          'aws-eu-north-1': 'aws-sct-builders-eu-north-1',
-                          'aws-eu-central-1': 'aws-sct-builders-eu-central-1',
-                          'aws-us-east-1' : 'aws-sct-builders-us-east-1-new',
+    def jenkins_labels = ['aws-eu-west-1': 'aws-sct-builders-eu-west-1-v2',
```
Comment (fruch) [inline vars/getJenkinsLabels.groovy:12]: make sure to add all the supported regions here.

File: sdcm/utils/sct_cmd_helpers.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
+import os
```
Comment (fruch) [inline sdcm/utils/sct_cmd_helpers.py:1]: add license header

**Review by soyacz — COMMENTED** [review]
Some minor comments, as it's hard to understand how it works just by looking at this change.

Topic looks very important and affects provisioning very much.
Can we have some session where you could describe the goal, problems and your solution?

File: jenkins-pipelines/artifacts-azure-image.jenkinsfile [hunk 3]
```diff
@@ -13,5 +13,4 @@ artifactsPipeline(
     timeout: [time: 145, unit: 'MINUTES'],
     post_behavior_db_nodes: 'destroy',
     builds_to_keep: '30',
-    ip_ssh_connections: 'public'
```
[the remark this answers is not in the document]
Comment (fruch) [reply jenkins-pipelines/artifacts-azure-image.jenkinsfile:16]: o.k., I'll revert this one.

in Azure there isn't anything like AWS vpc peering ?

File: sla_per_user_system_test.py [hunk 4]
```diff
@@ -60,7 +60,7 @@ def __init__(self, *args, **kwargs):
         self._es = ES()
 
     def prepare_schema(self):
-        self.prometheus_stats = PrometheusDBStats(host=self.monitors.nodes[0].public_ip_address)
+        self.prometheus_stats = PrometheusDBStats(host=self.monitors.nodes[0].external_address)
```
[the remark this answers is not in the document]
Comment (fruch) [reply sla_per_user_system_test.py:63]: as it's docsting says:
the communication address for usage between the test and the nodes

`external_address` can be ipv6/public/private base on the test configuration.
it's almost identical now to `ip_address`

File: jenkins-pipelines/artifacts-azure-image.jenkinsfile (hunk 3 above)
[the remark this answers is not in the document]
Comment (fruch) [reply jenkins-pipelines/artifacts-azure-image.jenkinsfile:16]: fixed, putting the configuration back

File: sdcm/utils/aws_region.py [hunk 5]
```diff
@@ -227,11 +237,52 @@ def create_sct_security_group(self):
                             }
                         ]
                     },
+                ]
+            )
+
+    def sct_test_security_group(self, test_id) -> EC2ServiceResource.SecurityGroup:
```
[the remark this answers is not in the document]
Comment (fruch) [reply sdcm/utils/aws_region.py]: changed

File: sdcm/utils/aws_region.py [hunk 6]
```diff
@@ -227,11 +237,52 @@ def create_sct_security_group(self):
...
+        LOGGER.debug("Found Security Groups: %s", security_groups)
+        existing_sgs = security_groups.get("SecurityGroups", [])
+        if len(existing_sgs) == 0:
+            return None
+        assert len(existing_sgs) == 1, \
+            f"More than 1 Security group with {name} found " \
+            f"in {self.region_name}: {existing_sgs}!"
+        return self.resource.SecurityGroup(existing_sgs[0]["GroupId"])  # pylint: disable=no-member
+
+    def create_sct_test_security_group(self, test_id: str):
```
[the remark this answers is not in the document]
Comment (fruch) [reply sdcm/utils/aws_region.py]: changed

File: requirements.txt [hunk 7]
```diff
@@ -45,9 +45,9 @@ azure-common==1.1.28 \
     #   azure-mgmt-resource
     #   azure-mgmt-resourcegraph
     #   azure-mgmt-subscription
-azure-core==1.23.1 \
-    --hash=sha256:28a01dfbaf0a6812c4e2a82d1642ea30956a9739f25bc77c9b23b91f4ea68f0f \
-    --hash=sha256:c3e8a9a3ec9d89f59b5d5b2f19d19a30d76a5b5c0cee3788ecad3cb72b9bd028
+azure-core==1.24.0 \
```
[the remark this answers is not in the document]
Comment (fruch) [reply requirements.txt:48]: @soyacz  test passed, is that good enough for saying azure api are working o.k. ?

[the remark this answers is not in the document]
Comment (soyacz) [reply requirements.txt:48]: yes. Thanks!

**Comment by soyacz** [comment]
Question: What happens if there are the same hostnames in 2 vpc's (I think this is our case as I didn't see region prefix in node names)? Won't they clash somehow?

**Comment by fruch** [comment]
> Question: What happens if there are the same hostnames in 2 vpc's (I think this is our case as I didn't see region prefix in node names)? Won't they clash somehow?

in AWS the DNS names are not based on the instance name, they are base on the public address if you have any:
```
ec2-52-18-93-225.eu-west-1.compute.amazonaws.com
```

the internal DNS also is based on the private ip:
```
ip-10-0-3-173.eu-west-1.compute.internal
```

VPC peering can't have collision ip addressed, you can't peer two VPCs that share the same CIDR.
hence the why this PR setup the VPC in a way that each region get a unique CIDR that doesn't overlap:
```
10.[region_index].0.0/16
````

So no clashing DNS, and no clashing ip addresses.
(but the need to build a whole new VPCs cause the current ones, all using 10.0.0.0/16)

You might want to consider that provisioning the regions in Azure

**Comment by fgelcer** [comment]
> So no clashing DNS, and no clashing ip addresses. (but the need to build a whole new VPCs cause the current ones, all using 10.0.0.0/16)
> 

so, how will work the multi-DC setup? will it be supported by current configuration?

**Comment by fruch** [comment]
> > So no clashing DNS, and no clashing ip addresses. (but the need to build a whole new VPCs cause the current ones, all using 10.0.0.0/16)
> 
> so, how will work the multi-DC setup? will it be supported by current configuration?

@fgelcer I'm not sure I'm following the question ? but
multi-dc in AWS would work, with this new configuration, but only with private addresses.
current configuration prior to this change multi-dc works only with public addresses.
this PR creates a whole new VPC and all AWS related permanent resources, hence it doesn't affect the current configuration (that we should keep until all the branches using it would be deprecated, or we'll take a call to backport those changes to those branches)

**Comment by fgelcer** [comment]
> > > So no clashing DNS, and no clashing ip addresses. (but the need to build a whole new VPCs cause the current ones, all using 10.0.0.0/16)
> > 
> > 
> > so, how will work the multi-DC setup? will it be supported by current configuration?
> 
> @fgelcer I'm not sure I'm following the question ? but multi-dc in AWS would work, with this new configuration, but only with private addresses. current configuration prior to this change multi-dc works only with public addresses. this PR creates a whole new VPC and all AWS related permanent resources, hence it doesn't affect the current configuration (that we should keep until all the branches using it would be deprecated, or we'll take a call to backport those changes to those branches)

great, another question, what about IPv6? will it work, or will it require any sort of enabling ports for it?

**Comment by fruch** [comment]
> > > > So no clashing DNS, and no clashing ip addresses. (but the need to build a whole new VPCs cause the current ones, all using 10.0.0.0/16)
> > > 
> > > 
> > > so, how will work the multi-DC setup? will it be supported by current configuration?
> > 
> > 
> > @fgelcer I'm not sure I'm following the question ? but multi-dc in AWS would work, with this new configuration, but only with private addresses. current configuration prior to this change multi-dc works only with public addresses. this PR creates a whole new VPC and all AWS related permanent resources, hence it doesn't affect the current configuration (that we should keep until all the branches using it would be deprecated, or we'll take a call to backport those changes to those branches)
> 
> great, another question, what about IPv6? will it work, or will it require any sort of enabling ports for it?

Didn't tested multi-dc ipv6 cases (I think maybe only the manger ipv6 is multi-dc ? or it's not multi-dc), but
the ipv6 ranges are configured in the SGs and in the route tables, so in theory it could work.

**Comment by fruch** [comment]
> > great, another question, what about IPv6? will it work, or will it require any sort of enabling ports for it?
> 
> Didn't tested multi-dc ipv6 cases (I think maybe only the manger ipv6 is multi-dc ? or it's not multi-dc), but the ipv6 ranges are configured in the SGs and in the route tables, so in theory it could work.


seems that `ping6` is working between the regions, tested between this setup jenkins workers:
```
jenkins@ip-10-4-0-239:~$ ping6  2a05:d016:38c:4a00:949d:fc47:3878:b138
PING 2a05:d016:38c:4a00:949d:fc47:3878:b138(2a05:d016:38c:4a00:949d:fc47:3878:b138) 56 data bytes
64 bytes from 2a05:d016:38c:4a00:949d:fc47:3878:b138: icmp_seq=1 ttl=64 time=42.2 ms
64 bytes from 2a05:d016:38c:4a00:949d:fc47:3878:b138: icmp_seq=2 ttl=64 time=42.2 ms
64 bytes from 2a05:d016:38c:4a00:949d:fc47:3878:b138: icmp_seq=3 ttl=64 time=42.2 ms
64 bytes from 2a05:d016:38c:4a00:949d:fc47:3878:b138: icmp_seq=4 ttl=64 time=42.3 ms
```

I should expected other problem, but I'll run the manager ipv6 case, nevertheless to see if it's not broken.

File: sdcm/utils/aws_region.py [hunk 8]
```diff
@@ -227,11 +237,52 @@ def create_sct_security_group(self):
...
+        LOGGER.debug("Found Security Groups: %s", security_groups)
+        existing_sgs = security_groups.get("SecurityGroups", [])
+        if len(existing_sgs) == 0:
+            return None
+        assert len(existing_sgs) == 1, \
+            f"More than 1 Security group with {name} found " \
+            f"in {self.region_name}: {existing_sgs}!"
+        return self.resource.SecurityGroup(existing_sgs[0]["GroupId"])  # pylint: disable=no-member
+
+    def provide_sct_test_security_group(self, test_id: str):
```
Comment (soyacz) [inline sdcm/utils/aws_region.py:259]: This `test_security_group` is about opening all the related ports to public?
Line of comment would be helpful.

Cleanup resources cleans this sg?

**Review by soyacz — COMMENTED** [review]
Generally some things could be more 'object oriented' in a way we could reuse or replace classes for different cloud providers. Things like ssh_login_info, instance, connect_vm are simple dicts and code user must remember about all the keys in these dicts to be able to use it.
But I'm not pushing to refactor now, if it's done and works, it's better to refactor when we want to introduce equivalent of VPC peering in other cloud providers - then some unification will be necessary.

Can you provide some links to example test executions?

File: sdcm/utils/aws_region.py (hunk 8 above)
Comment (fruch) [reply sdcm/utils/aws_region.py:259]: yes, cleanup does clean it up.

File: jenkins-pipelines/manager/centos-manager-sanity-ipv6.jenkinsfile [hunk 9]
```diff
@@ -6,9 +6,9 @@ def lib = library identifier: 'sct@snapshot', retriever: legacySCM(scm)
 managerPipeline(
     backend: 'aws',
     ip_ssh_connections: 'ipv6',
-    region: 'eu-west-1',
+    region: '''["us-east-1", "us-west-2"]''',
     test_name: 'mgmt_cli_test.MgmtCliTest.test_manager_sanity',
-    test_config: 'test-cases/manager/manager-regression-ipv6.yaml',
+    test_config: 'test-cases/manager/manager-regression-multiDC-set-distro.yaml',
```
[the remark this answers is not in the document]
Comment (fruch) [reply jenkins-pipelines/manager/centos-manager-sanity-ipv6.jenkinsfile:11]: it was for testing, but I'm leaving this change for raya's team (owners of the manger tests)

File: sct.py [hunk 10]
```diff
@@ -1300,16 +1284,28 @@ def create_test_release_jobs_enterprise(branch, username, password, sct_branch,
 
 @cli.command("prepare-region", help="Configure all required resources for SCT runs in selected cloud region")
 @cloud_provider_option
-@click.option("-r", "--region", required=True, type=CloudRegion(), help="Cloud region")
-def prepare_region(cloud_provider, region):
+@click.option("-r", "--regions", type=CloudRegion(), help="Cloud region", multiple=True)
+def prepare_region(cloud_provider, regions):
```
[the remark this answers is not in the document]
Comment (fruch) [reply sct.py:1288]: renamed

File: sct_ssh.py [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+
+# pylint: disable=too-many-arguments
+
+@click.command("ssh", help="Connect to any SCT machine on AWS")
+@click.option("-u", "--user", default=getpass.getuser(),
+              show_default=True,
+              help="User to search for (RunByUser tag)")
+@click.option("-t", "--test-id", default=None, help="test id to search for")
+@click.option("-r", "--region", default=None, help="region to use, default search across all regions")
+@click.option("-P", "--public-ip", is_flag=True, show_default=True, default=False,
```
[the remark this answers is not in the document]
Comment (fruch) [reply sct_ssh.py]: changed

File: sct_ssh.py [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+
+
+def select_instance(region: str = None, **tags) -> dict | None:
+    user = tags.get('user')
+    test_id = tags.get('test_id')
+    tags = {}
+    if user:
+        tags.update({"RunByUser": user})
+    if test_id:
+        tags.update({"TestId": test_id})
```
[the remark this answers is not in the document]
Comment (fruch) [reply sct_ssh.py:86]: Removed the default user

File: sct_ssh.py [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+from questionary import Choice
+
+from sdcm.utils.common import list_instances_aws, get_free_port
+from sdcm.utils.aws_region import AwsRegion
+
+
+def get_region(instance: dict) -> str:
+    return instance.get('Placement').get('AvailabilityZone')[:-1]
+
+
```
[the remark this answers is not in the document]
Comment (fruch) [reply sct_ssh.py:30]: why ? each instance passed to this, might be in a different region. 
anyhow it's dict access, not a call to AWS. I don't see reason to cache it.

File: sdcm/utils/aws_region.py [hunk 14]
```diff
@@ -20,25 +20,35 @@
...
-    SCT_ROUTE_TABLE_NAME = "SCT-rt"
+    SCT_VPC_NAME = "SCT-2-vpc"
+    SCT_VPC_CIDR_TMPL = "10.{}.0.0/16"
+    SCT_SECURITY_GROUP_NAME = "SCT-2-sg"
+    SCT_TEST_SECURITY_GROUP_NAME_TMPL = "SCT-2-sg-{}"
+    SCT_SUBNET_NAME = "SCT-2-subnet-{availability_zone}"
+    SCT_INTERNET_GATEWAY_NAME = "SCT-2-igw"
+    SCT_ROUTE_TABLE_NAME = "SCT-2-rt"
     SCT_KEY_PAIR_NAME = "scylla-qa-ec2"  # TODO: change legacy name to sct-keypair-aws
+    SCT_BUILDR_SSH_GROUP_NAME = 'SCT-builder-ssh-sg'
```
[the remark this answers is not in the document]
Comment (fruch) [reply sdcm/utils/aws_region.py:38]: the name have both the group it's relates to, i.e. the builder, and the protocol i.e ssh.

how come the action "open" is better the the group it's related to ?

File: sdcm/utils/aws_region.py (hunk 8 above)
Comment (fruch) [reply sdcm/utils/aws_region.py:259]: Also added a docstring to this function, explaining when this should be used.

File: sct_ssh.py [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+@click.command("ssh", help="Connect to any SCT machine on AWS")
+@click.option("-u", "--user", default=getpass.getuser(),
+              show_default=True,
+              help="User to search for (RunByUser tag)")
+@click.option("-t", "--test-id", default=None, help="test id to search for")
+@click.option("-r", "--region", default=None, help="region to use, default search across all regions")
+@click.option("-P", "--public-ip", is_flag=True, show_default=True, default=False,
+              help="Force usage of public address")
+def ssh(user, test_id, region, public_ip):
```
[the remark this answers is not in the document]
Comment (fruch) [reply sct_ssh.py]: Added the support for it

**Comment by fruch** [comment]
need to rebuild the docker image

**Comment by fruch** [comment]
@roydahan, I think we are ready to merge this one, I doesn't think it's gonna impact our backports too much. it's touching areas we don't usually touch and change for bug fixes or such.

File: sdcm/utils/aws_region.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (soyacz) [reply sdcm/utils/aws_region.py:38]: I think I meant that this group was not used for builders only but also for sct-runners and in that sense `BUILDR` is confusing.

**Comment by fruch** [comment]
> @roydahan, I think we are ready to merge this one, I doesn't think it's gonna impact our backports too much. it's touching areas we don't usually touch and change for bug fixes or such.


@roydahan now that we finished(released) 5.0/2022.1, I think we are ready to merge this one.

**Comment by roydahan** [comment]
> > @roydahan, I think we are ready to merge this one, I doesn't think it's gonna impact our backports too much. it's touching areas we don't usually touch and change for bug fixes or such.
> 
> @roydahan now that we finished(released) 5.0/2022.1, I think we are ready to merge this one.

Ok, but did it pass review?
I see all review statuses as not reviewed.

**Comment by fruch** [comment]
> > > @roydahan, I think we are ready to merge this one, I doesn't think it's gonna impact our backports too much. it's touching areas we don't usually touch and change for bug fixes or such.
> > 
> > 
> > @roydahan now that we finished(released) 5.0/2022.1, I think we are ready to merge this one.
> 
> Ok, but did it pass review? I see all review statuses as not reviewed.

@soyacz did a review, the only standing left comment is on the name SCT_BUILDR_SSH_GROUP_NAME.
If we change it, I'll need to go over manually and change in all regions.

**Comment by fruch** [comment]
And also now I have conflict to fix, and rebuild docker image

**Comment by soyacz** [comment]
@fruch do you have test job for this PR?

**Comment by fruch** [comment]
> @fruch do you have test job for this PR?

I've run most of the testing in this one:
https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/longevity-alternator-3h-multidc-test/19/

**Comment by soyacz** [comment]
> > @fruch do you have test job for this PR?
> 
> I've run most of the testing in this one: https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/longevity-alternator-3h-multidc-test/19/

Looks like collect-logs is using public ip's for connections and failing. Is collect-logs working?

**Comment by fruch** [comment]
> > > > @fruch do you have test job for this PR?
> > > 
> > > 
> > > I've run most of the testing in this one: https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/longevity-alternator-3h-multidc-test/19/
> > 
> > 
> > Looks like collect-logs is using public ip's for connections and failing. Is collect-logs working?
> 
> I'll try it again, but the collect should be now working with local ip address, maybe I've missed a spot i'll look again

I've handled it in the code and it should be using public addresses, 
I'm running this one again
https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/longevity-alternator-3h-multidc-test/20/

**Review by soyacz — APPROVED** [review]
LGTM

**Comment by fruch** [comment]
seems like the provision test with sct-runner isn't working. investigating...

**Comment by fruch** [comment]
> seems like the provision test with sct-runner isn't working. investigating...

weird, https://jenkins.scylladb.com/job/sct-github-PRs-scan/job/scylla-cluster-tests/job/PR-4742/43/ starts fine...
I'll let it finish

**Review by roydahan — DISMISSED** [review]

**Pull request merged**